src/Components/test/Ignitor/RenderBatchReader.cs

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.AspNetCore.Components.RenderTree;
+
+namespace Ignitor
+{
+    public static class RenderBatchReader
+    {
+        public static RenderBatch Read(ReadOnlySpan<byte> data)
+        {
+            var sections = Sections.Parse(data);
+            var strings = ReadStringTable(data, sections.GetStringTableIndexes(data));
+            var diffs = ReadUpdatedComponents(data, sections.GetUpdatedComponentIndexes(data), strings);
+            var frames = ReadReferenceFrames(sections.GetReferenceFrameData(data), strings);
+            var disposedComponentIds = ReadDisposedComponentIds(data);
+            var disposedEventHandlerIds = ReadDisposedEventHandlerIds(data);
+            return new RenderBatch(diffs, frames, disposedComponentIds, disposedEventHandlerIds);
+        }
+
+        private static string[] ReadStringTable(ReadOnlySpan<byte> data, ReadOnlySpan<byte> indexes)
+        {
+            var result = new string[indexes.Length / 4];
+
+            for (var i = 0; i < indexes.Length; i += 4)
+            {
+                var index = BitConverter.ToInt32(indexes.Slice(i, 4));
+
+                // The string table entries are all length-prefixed UTF8 blobs
+                var length = (int)ReadUnsignedLEB128(data, index, out var numLEB128Bytes);
+                var value = Encoding.UTF8.GetString(data.Slice(index + numLEB128Bytes, length));
+                result[i / 4] = value;
+            }
+
+            return result;
+        }
+
+        private static ArrayRange<RenderTreeDiff> ReadUpdatedComponents(ReadOnlySpan<byte> data, ReadOnlySpan<byte> indexes, string[] strings)
+        {
+            var result = new RenderTreeDiff[indexes.Length / 4];
+
+            for (var i = 0; i < indexes.Length; i += 4)
+            {
+                var index = BitConverter.ToInt32(indexes.Slice(i, 4));
```

Comment:
I think some of these might have benefited from using a `SequenceReader`. That way you wouldn't have to manually track the indices. But I wouldn't bother changing the codez at this point.

Reply:
Left as-is for now.